crates/runtime/src/backtrace.rs

```diff
@@ -0,0 +1,146 @@
+//! Backtrace object and utilities.
+
+use crate::jit_frame_registry;
+
+/// Information about backtrace frame.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct BacktraceFrame {
+    pc: usize,
+}
+
+impl Default for BacktraceFrame {
+    fn default() -> Self {
+        Self { pc: 0 }
+    }
+}
+
+impl BacktraceFrame {
+    /// Current PC or IP pointer for the frame.
+    pub fn pc(&self) -> usize {
+        self.pc
+    }
+    /// Additinal frame information.
+    pub fn tag(&self) -> Option<jit_frame_registry::JITFrameTag> {
+        jit_frame_registry::find(self.pc)
+    }
+}
+
+const BACKTRACE_LIMIT: usize = 32;
+
+/// Backtrace during WebAssembly trap.
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub struct Backtrace {
+    len: usize,
+    frames: [BacktraceFrame; BACKTRACE_LIMIT],
+}
+
+impl Backtrace {
+    fn new() -> Self {
+        Self {
+            len: 0,
+            frames: [Default::default(); BACKTRACE_LIMIT],
+        }
+    }
+    fn full(&self) -> bool {
+        self.len >= BACKTRACE_LIMIT
+    }
+    fn push(&mut self, frame: BacktraceFrame) {
+        assert!(self.len < BACKTRACE_LIMIT);
+        self.frames[self.len] = frame;
+        self.len += 1;
+    }
+    /// Amount of the backtrace frames.
+    pub fn len(&self) -> usize {
+        self.len
+    }
+}
+
+impl std::ops::Index<usize> for Backtrace {
+    type Output = BacktraceFrame;
+    fn index(&self, index: usize) -> &Self::Output {
+        assert!(index < self.len);
+        &self.frames[index]
+    }
+}
+
+impl std::fmt::Debug for Backtrace {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(f, "Backtrace![")?;
+        for i in 0..self.len() {
+            let frame = &self.frames[i];
+            writeln!(f, "  {:x}: {:?}", frame.pc(), frame.tag())?;
+        }
+        write!(f, "]")?;
+        Ok(())
+    }
+}
+
+#[cfg(not(all(target_os = "windows", target_arch = "x86_64")))]
+fn capture_stack<F>(mut f: F)
+where
+    F: FnMut(usize) -> bool,
+{
+    use backtrace::trace;
+    trace(|frame| {
+        let pc = frame.ip() as usize;
+        f(pc)
+    });
+}
+
+#[cfg(all(target_os = "windows", target_arch = "x86_64"))]
```

Comment:
I was running this locally on MSVC, and the test added in this PR actually passes without this custom implementation. I suspect though it doesn't work as-is as I might think? Could a test be added which requires this implementation on Windows?

Reply:
Yeah I checked this PR out locally and the test passed, although I didn't dig in too closely. I do think a more extensive backtrace test would be good, but we could perhaps avoid the addition of this windows-specific routine until an added test fails?

Reply:
My concern with that approach is that Wasm stack traces on Windows might not be reliable without this implementation.

I'd rather add a test case that verifies this code works as an improvement over `backtrace` (i.e. a test case that fails if we use `backtrace` but works with this implementation) as part of this PR.

Reply:
I guess that it's probably best if we have additional test cases anyway.

Some suggestions:

* Simple recursion (easy to test the current 32 frame limit).
* Calls to functions that allocate stack space.
* Complex traces containing multiple wasm-to-host and host-to-wasm transitions.

Etc.

Reply:
Add tests that reproduces issue from #751